tests/sentry/api/endpoints/test_organization_details.py

```diff
@@ -764,11 +766,69 @@ def test_update_name_with_mapping(self):
         org = Organization.objects.get(id=organization_id)
         assert org.name == "SaNtRy"
 
+        with outbox_runner():
+            pass
+
         with exempt_from_silo_limits():
             assert OrganizationMapping.objects.filter(
                 organization_id=organization_id, name="SaNtRy"
             ).exists()
 
+    def test_update_slug(self):
+        organization_mapping = OrganizationMapping.objects.get(organization_id=self.organization.id)
+        assert organization_mapping.slug == self.organization.slug
+
+        desired_slug = "new-santry"
+        self.get_success_response(self.organization.slug, slug=desired_slug)
+        self.organization.refresh_from_db()
+        assert self.organization.slug == desired_slug
+
+        # Ensure that the organization update has been flushed
+        with outbox_runner():
+            pass
+
+        organization_mapping.refresh_from_db()
+        assert organization_mapping.slug == desired_slug
+
+    def test_update_slug_with_temporary_rename_collision(self):
+        desired_slug = "taken"
+        previous_slug = self.organization.slug
+        org_with_colliding_slug = self.create_organization(
+            slug=desired_slug, name="collision-imminent"
+        )
+
+        # Drain the initial slug creation to ensure a mapping exists for the new org
+        Organization.outbox_for_update(org_id=org_with_colliding_slug.id).drain_shard()
+        colliding_org_mapping = OrganizationMapping.objects.get(
+            organization_id=org_with_colliding_slug.id
+        )
+        assert colliding_org_mapping.slug == desired_slug
+
+        # Queue a slug update but don't drain the shard yet to ensure a temporary collision happens
+        org_with_colliding_slug.slug = "unique-slug"
+        org_with_colliding_slug.save()
+
+        self.get_success_response(self.organization.slug, slug=desired_slug)
+        self.organization.refresh_from_db()
+        assert self.organization.slug == desired_slug
+
+        # Ensure that the organization update has been flushed, but it collides when attempting an upsert
+        with pytest.raises(IntegrityError):
+            Organization.outbox_for_update(org_id=self.organization.id).drain_shard()
```

Comment:
With both organization_id and slug being unique, could we end up with unsolvable cycles? The scenario I thought of involves two organizations that want to switch their slugs.

1. Org id=123 slug=first - Changes their slug to 'taken'
2. Org id=234 slug=taken - Changes their slug to 'first'

If synchronous outbox delivery fails for both of these operations, then both orgs could end up with stuck outbox messages as neither message can be processed in isolation 🤔 This will be exceptionally rare and doesn't have to block merging these changes but we should think about how to resolve this state more.

Reply:
Hmm I'm not sure I see how this would be possible since we're relying on the Organization table's slug unique key constraint to guard against this. There would have to be some intermediary, non-colliding slug Org 1 would first have to select before Org 2 would be allowed to change its slug to `first`, then allowing Org 1 to take Org 2's old slug.

I suppose if the order were:
1. Org id=123, slug=first -> changes their slug to 'non-colliding-slug'
2. Org id=234, slug=taken - Changes their slug to 'first'
3. Org id=123, slug='non-colliding-slug' - changes their slug to taken
and then we process the outbox this could occur?

The thing we'll have to be cognizant of is how we coalesce these org update outbox messages to ensure that it resolves each slug change correctly in order and doesn't deadlock our slug swap essentially. This also applies to the new proposed design where the control-silo is the source of truth so it's something @corps and I are going to address in our mini tech spec.